action/protocol/staking/candidate_statemanager.go

```diff
@@ -128,26 +162,193 @@ func (csm *candSM) Upsert(d *Candidate) error {
+	// Add index inside bucket
+	bucket.Index = index
+	if _, err := csm.PutState(
+		bucket,
+		protocol.NamespaceOption(StakingNameSpace),
+		protocol.KeyOption(bucketKey(index))); err != nil {
+		return 0, err
+	}
+	tc.count++
+	_, err := csm.PutState(
+		&tc,
+		protocol.NamespaceOption(StakingNameSpace),
+		protocol.KeyOption(TotalBucketKey))
+	return index, err
+}
+
+func (csm *candSM) delBucket(index uint64) error {
+	_, err := csm.DelState(
+		protocol.NamespaceOption(StakingNameSpace),
+		protocol.KeyOption(bucketKey(index)))
+	return err
+}
+
+func (csm *candSM) putBucketAndIndex(bucket *VoteBucket) (uint64, error) {
+	index, err := csm.putBucket(bucket)
+	if err != nil {
+		return 0, errors.Wrap(err, "failed to put bucket")
+	}
+
+	if err := putVoterBucketIndex(csm, bucket.Owner, index); err != nil {
+		return 0, errors.Wrap(err, "failed to put bucket index")
+	}
+
+	if err := putCandBucketIndex(csm, bucket.Candidate, index); err != nil {
+		return 0, errors.Wrap(err, "failed to put candidate index")
+	}
+	return index, nil
+}
+
+func (csm *candSM) delBucketAndIndex(owner, cand address.Address, index uint64) error {
+	if err := csm.delBucket(index); err != nil {
+		return errors.Wrap(err, "failed to delete bucket")
+	}
+
+	if err := delVoterBucketIndex(csm, owner, index); err != nil {
+		return errors.Wrap(err, "failed to delete bucket index")
+	}
+
+	if err := delCandBucketIndex(csm, cand, index); err != nil {
+		return errors.Wrap(err, "failed to delete candidate index")
+	}
+	return nil
+}
+
+func (csm *candSM) putBucketIndex(key []byte, index uint64) error {
+	var bis BucketIndices
+	if _, err := csm.State(
+		&bis,
+		protocol.NamespaceOption(StakingNameSpace),
+		protocol.KeyOption(key)); err != nil && errors.Cause(err) != state.ErrStateNotExist {
+		return err
+	}
+	bis.addBucketIndex(index)
+	_, err := csm.PutState(
+		&bis,
+		protocol.NamespaceOption(StakingNameSpace),
+		protocol.KeyOption(key))
+	return err
+}
+
+func (csm *candSM) delBucketIndex(key []byte, index uint64) error {
+	var bis BucketIndices
+	if _, err := csm.State(
+		&bis,
+		protocol.NamespaceOption(StakingNameSpace),
+		protocol.KeyOption(key)); err != nil {
+		return err
+	}
+	bis.deleteBucketIndex(index)
+
+	var err error
+	if len(bis) == 0 {
+		_, err = csm.DelState(
+			protocol.NamespaceOption(StakingNameSpace),
+			protocol.KeyOption(key))
+	} else {
+		_, err = csm.PutState(
+			&bis,
+			protocol.NamespaceOption(StakingNameSpace),
+			protocol.KeyOption(key))
+	}
+	return err
+}
+
+func (csm *candSM) putCandidate(d *Candidate) error {
+	_, err := csm.PutState(d, protocol.NamespaceOption(CandidateNameSpace), protocol.KeyOption(d.Owner.Bytes()))
+	return err
+}
+
+func (csm *candSM) delCandidate(name address.Address) error {
+	_, err := csm.DelState(protocol.NamespaceOption(CandidateNameSpace), protocol.KeyOption(name.Bytes()))
+	return err
+}
+
+// settleAccount deposits gas fee and updates caller's nonce
+func (csm *candSM) settleAction(
```

Comment:
this should be staking protocol's func, not csm's
it's kinda counter-intuitive that csm needs to take p *Protocol as input

Reply:
ok , i will revert it

Reply:
reverted @dustinxie